Unbreak 'use NativeCall' on JVM

usev6 opened the pull request:
... by not assigning nqp::hash at compile time. This blows up on the
JVM backend -- see https://github.com/rakudo/rakudo/issues/3043.

usev6 commented:
I'm not totally sure, but I was under the impression, that the JVM backend *did* support calls into native C libraries (via ```com.sun.jna```) -- and that the code just bitrotted. If I'm not mistaken the first commit in that area was https://github.com/Raku/nqp/commit/cc25bbc39d -- that's even before MoarVM did know about NativeCall. Maybe someone with more knowledge of the JVM backend could shed some light on this?

But anyway, your point still stands: We'd remove an optimization to fix something that no-one uses. Do you have a rough guess how much we'd lose without the optimization?

Of course, fixing the underlying problem with ```my constant $foo = nqp::hash()``` on the JVM backend would be best. But I'm afraid that's above my paygrade.

niner commented:
I'm pretty sure that we could get NativeCall back on JVM quite easily. I just 
forgot about JVM when introducing the nativeinvoke op. While implementing that 
op would be pretty hard, there's no hard requirement to use it.

We'd just need to keep the generic function body instead of replacing it here:
https://github.com/rakudo/rakudo/blob/master/lib/NativeCall.rakumod#L315

usev6 commented:
Thanks for the pointer. I'll try to look into this -- unless you want to do that.

I think, we'll also need a ```nativecallinvoke``` op, but that's probably just a copy of the current ```nativecall``` op.

usev6 commented:
I've opened a PR to add the op ```nativecallinvoke``` (as a copy of ```nativecall```): https://github.com/Raku/nqp/pull/708

> We'd just need to keep the generic function body instead of replacing it here:
> https://github.com/rakudo/rakudo/blob/master/lib/NativeCall.rakumod#L315

I looked a while, but I don't see/understand how this would help. As far as I can see we'd have to avoid calling ```!create-jit-compiled-function-body``` -- because that method uses ```nativeinvoke```.

For experimentation I've just changed https://github.com/rakudo/rakudo/blob/ac9f40d871/lib/NativeCall.rakumod#L541 to call ```!create-function-body``` a second time.

This allowed some progress, but showed another problem: The JVM backend doesn't know about ```$!call```, introduced with https://github.com/rakudo/rakudo/commit/9ef7a7e9dd.

With another workaround added, a couple of test files started to pass. But there seems to be a severe problem if an extra block is added around the definition of a native sub:

```
$ ./rakudo-j -e 'use NativeCall; { sub abs(uint32 $n --> uint32) is native { * }; say abs(-42) }'
java.lang.NullPointerException
  in sub abs at /home/christian/perl6/perl6-roast-data/rakudo.jvm/lib/NativeCall.rakumod (NativeCall) line 597
  in block <unit> at -e line 1
```

It works without the extra block:
```
$ ./rakudo-j -e 'use NativeCall; sub abs(uint32 $n --> uint32) is native { * }; say abs(-42)'
42
```

usev6 commented:
> Of course, fixing the underlying problem with my constant $foo = nqp::hash() on the JVM backend would be best. But I'm afraid that's above my paygrade.

Well, again stubbornness seems to have paid off. I think I've found the underlying problem: https://github.com/Raku/nqp/pull/717

usev6 commented:
This PR is no longer needed. Closing.
